---
title: "PLX5622 Diet Mice scRNAseq"
subtitle: "All Clusters Pass 1 QC and DE"
author: "Kennedi Todd"
date: "11/14/2023"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
library(ComplexUpset) 
library(dplyr)        # ungroup()
library(ggrepel)      # geom_text_repel()
library(ggtree)       # BuildClusterTree()
library(gridExtra)    # grid.arrange()
library(gtools)       # smartbind()
library(parallel)     # detectCores()
library(plotly)       # plot_ly()
library(Seurat)       # DimPlot()
library(SeuratObject)
options(Seurat.object.assay.version = "v4")
library(tidyr)        # %>%
library(UpSetR)       # fromList()

# work in parallel
options(mc.cores = detectCores() - 1)
```

## Variables and functions
```{r set_variables_and_functions}
# variables
sample_order <- c("E3CF1","E3CF2","E3CM1","E3CM2","E3PF1","E3PF2","E3PM1","E3PM2",
                  "E4CF1","E4CF2","E4CM1","E4CM2","E4PF1","E4PF2","E4PM1","E4PM2")
sample_colors <- c("#9e0909","#f75959","#f0b402","#f5d67a",
                   "#fbff0a","#fafaa7","#1d8c0e","#63e851",
                   "#0271f0","#9cc3f0","#6c39f7","#c589fa",
                   "#ed2af7","#f5c1f7","#755410","#b5aa82")
group_order <- c("E3CM","E3PM","E4CM","E4PM",
                 "E3CF","E3PF","E4CF","E4PF")
group_colors <- c("#B89B74","#725E47","#B15928","#F7590B",
                  "#A6CEE3","#34B1E8","#ED8F47","#F4CC25")
group2_order <- c("E3C","E3P","E4C","E4P")
group2_colors <- c("#f75959","#fbff0a","#63e851","#0271f0")
isoform_order <- c("E4","E3")
isoform_colors <- c("darkgray","cornflowerblue")
sex_order <- c("Male","Female")
sex_colors <- c("green","purple")
diet_order <- c("control","PLX5622")
diet_color <- c("coral","cyan")

# single cell functions
source("../../../functions/Kennedi_single_cell_functions_v4.R")

# save function
saveToPDF <- function(...) {
    d = dev.copy(pdf,...)
    dev.off(d)
}
```

## Load data
```{r read_pass1_annotated_obj}
mouse.annotated <- readRDS("../../rObjects/pass1_annotated.rds")
```

# QC
## 2D UMAP
```{r 2D_annotated_umap}
# set colors
cluster_colors <- c("#B5B9BA",
                    "#3385BB",
                    "#40BBFF",
                    "#84BF96",
                    "#7F9D55",
                    "#F57C7C",
                    "#E42622",
                    "#FBB268",
                    "#FE8D19",
                    "#DE9E83",
                    "#A6CEE3",
                    "#9D7BBA",
                    "#977899")


# umap
umap <- DimPlot(object = mouse.annotated, 
        reduction = "umap",
        repel = TRUE,
        group.by = "annotated_clusters",
        cols = cluster_colors)
umap
```

## 3D UMAP
```{r 3D_annotated_UMAP}
embeddings <- mouse.annotated@reductions$umap@cell.embeddings
embeddings <- cbind(embeddings, as.character(mouse.annotated$annotated_clusters))
colnames(embeddings)[4] <- "annotated_clusters"
embeddings <- as.data.frame(embeddings)

three.dim <- plot_ly(embeddings,
                     x = ~UMAP_1, 
                     y = ~UMAP_2, 
                     z = ~UMAP_3, 
                     color = ~annotated_clusters, 
                     colors = cluster_colors,
                     size = 1) 
three.dim <- three.dim %>% add_markers() 
three.dim <- three.dim %>% layout(scene = list(xaxis = list(title = 'UMAP_1'), 
                                     yaxis = list(title = 'UMAP_2'), 
                                     zaxis = list(title = 'UMAP_3')))
three.dim
```

## Cluster markers
```{r cluster_markers_violin}
goi <- c("Pdgfrb","Acta2","Vwf","Cldn5","Pecam1","Flt4","Prox1","Lyve1","Ptprc",
         "Cd19","Ms4a1","Ighd","Cd3e","Trbc2","Il7r","Nkg7","Klrb1b","Klrb1c",
         "Gata3","Rora","Itgax","H2-Eb1","Ccr2","Ly6c2","Lyz2","Ly6g","Itgam",
         "Mrc1","Csf1r","Cd38","Mki67","Mcpt4","Ms4a2","Col1a2","Plp1")

v <- VlnPlot(mouse.annotated,
             features = goi,
             split.by = "annotated_clusters",
             flip = TRUE,
             stack = TRUE,
             cols = cluster_colors)
v
```

## Split UMAP
```{r split_umaps}
# Apoe isoform
umap1 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "isoform",
                 repel = TRUE,
                 cols = cluster_colors)
umap1

# sex
umap2 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "sex",
                 repel = TRUE,
                 cols = cluster_colors)
umap2

# sample
umap3 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "sample",
                 ncol = 4,
                 repel = TRUE,
                 cols = cluster_colors)
umap3

# phase
umap4 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "phase",
                 cols = cluster_colors,
                 repel = TRUE)
umap4

# mito.factor
umap5 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "mito.factor",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap5

# group
umap6 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "group",
                 cols = cluster_colors,
                 ncol = 4,
                 repel = TRUE)
umap6

# group2
umap7 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "group2",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap7

# diet
umap8 <- DimPlot(object = mouse.annotated, 
                 group.by = "annotated_clusters",
                 reduction = "umap",
                 split.by = "diet",
                 cols = cluster_colors,
                 ncol = 2,
                 repel = TRUE)
umap8
```

```{r, echo=FALSE, eval=FALSE}
umap1
path <- paste0("../../results/all_clusters_pass1/clustering_QC/")
saveToPDF(paste0(path, "split_isoform_umap.pdf"), width = 12, height = 6)
dev.off()

umap2
saveToPDF(paste0(path, "split_sex_umap.pdf"), width = 12, height = 6)
dev.off()

umap3
saveToPDF(paste0(path, "split_sample_umap.pdf"), width = 14, height = 10)
dev.off()

umap4
saveToPDF(paste0(path, "split_cell_cycle_phase_umap.pdf"), width = 14, height = 6)
dev.off()

umap5
saveToPDF(paste0(path, "split_mito_factor_umap.pdf"), width = 12, height = 8)
dev.off()

umap6
saveToPDF(paste0(path, "split_group_umap.pdf"), width = 16, height = 8)
dev.off()

umap7
saveToPDF(paste0(path, "split_group2_umap.pdf"), width = 12, height = 8)
dev.off()

umap8
saveToPDF(paste0(path, "split_diet_umap.pdf"), width = 12, height = 6)
dev.off()

remove(umap1,umap2,umap3,umap4,umap5,umap6,umap7)
```

## Heatmap UMAP
```{r feature_plots_annotated, warning=FALSE, message=FALSE}
# UMAP percent.mt
f1 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent.mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse.annotated, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent.ribo
f4 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent.ribo.protein") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell.complexity
f5 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "cell.complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5
```

```{r, echo=FALSE, eval=FALSE}
path <- "../../results/all_clusters_pass1/clustering_QC/heatmap_"
f1
saveToPDF(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
dev.off()

f2
saveToPDF(paste0(path, "nCount.pdf"), width = 8, height = 6)
dev.off()

f3
saveToPDF(paste0(path, "nFeature.pdf"), width = 8, height = 6)
dev.off()

f4
saveToPDF(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
dev.off()

f5
saveToPDF(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
dev.off()

remove(f1,f2,f3,f4,f5)
```

## Percent cells per cluster
```{r percent_cells}
# isoform
b1 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, isoform) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=isoform)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = isoform_colors) +
  ggtitle("Percentage of isoform per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sex) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sex_colors) +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sample) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sample)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = sample_colors) +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, phase) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, mito.factor) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=mito.factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito.factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group2) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = group2_colors) +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# diet
b8 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, diet) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=diet)) +
  theme_classic() +
  geom_col() +
  scale_fill_manual(values = diet_color) +
  ggtitle("Percentage of diet per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8
```

```{r,echo=FALSE,eval=FALSE}
# save
b1
path <- "../../results/all_clusters_pass1/clustering_QC/percent_isoform_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b2
path <- "../../results/all_clusters_pass1/clustering_QC/percent_sex_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b3
path <- "../../results/all_clusters_pass1/clustering_QC/percent_sample_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b4
path <- "../../results/all_clusters_pass1/clustering_QC/percent_phase_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b5
path <- "../../results/all_clusters_pass1/clustering_QC/percent_mito_factor_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()


# save
b6
path <- "../../results/all_clusters_pass1/clustering_QC/percent_group_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b7
path <- "../../results/all_clusters_pass1/clustering_QC/percent_group2_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# save
b8
path <- "../../results/all_clusters_pass1/clustering_QC/percent_diet_per_cluster"
saveToPDF(paste0(path, ".pdf"), width = 8, height = 6)
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8)
```

## Cluster tree
```{r cluster_tree_indv_annotated, message=FALSE, warning=FALSE}
Idents(mouse.annotated) <- mouse.annotated$annotated_clusters
mouse.annotated <- BuildClusterTree(object = mouse.annotated,
                                     dims = 1:15,
                                     reorder = FALSE,
                                     reorder.numeric = FALSE)

tree <- mouse.annotated@tools$BuildClusterTree
tree$tip.label <- paste0(tree$tip.label)

p <- ggtree::ggtree(tree, aes(x, y)) +
  scale_y_reverse() +
  ggtree::geom_tree() +
  ggtree::theme_tree() +
  ggtree::geom_tiplab(offset = 1) +
  ggtree::geom_tippoint(color = cluster_colors[1:length(tree$tip.label)], shape = 16, size = 5) +
  coord_cartesian(clip = 'off') +
  theme(plot.margin = unit(c(0,2.5,0,0), 'cm'))
p
```

```{r,echo=FALSE,eval=FALSE}
p
path <- paste0("../../results/all_clusters_pass1/clustering_QC/cluster_tree")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 10)
dev.off()
```

## Cells per sample
```{r}
data <- as.data.frame(table(mouse.annotated$sample))
colnames(data) <- c("sample","frequency")

ncells <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,8000, by = 1000), limits = c(0,8000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r,echo=FALSE,eval=FALSE}
ncells
path <- "../../results/all_clusters_pass1/clustering_QC/cells_per_sample"
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
dev.off()
```

# Differential expression
## E3 PLX5622 male vs E3 control male within each cluster
```{r E3PM_vs_E3CM, warning=FALSE, message=FALSE, eval=FALSE}
# intitialize variables
cell_types <- levels(mouse.annotated$annotated_clusters)
e3.male.df <- data.frame()

# loop through clusters
for (i in cell_types) {
  print(i)
  cluster <- subset(mouse.annotated, annotated_clusters == i)
  Idents(cluster) <- cluster$group
  markers <- FindMarkers(object = cluster,
                         ident.1 = "E3PM",
                         ident.2 = "E3CM",
                         only.pos = FALSE, # default
                         min.pct = 0.10,   # default
                         test.use = "MAST",
                         verbose = TRUE,
                         assay = "RNA")
  if(nrow(markers) == 0) {
    next
  }
  markers$cluster <- i
  markers$gene <- rownames(markers)
  e3.male.df <- rbind(e3.male.df, markers)
}

# reformat table
colnames(e3.male.df)[c(3,4)] <- c("percent_E3PM","percent_E3CM")
rownames(e3.male.df) <- 1:nrow(e3.male.df)
e3.male.df$percent_difference <- abs(e3.male.df$percent_E3PM - e3.male.df$percent_E3CM)
e3.male.df <- e3.male.df[,c(6,7,1,5,2,3,4,8)]

# write table
write.table(e3.male.df, "../../results/all_clusters_pass1/DEGs/DEG_tables/E3PM_vs_E3CM_DEGs.tsv", 
            sep = "\t", quote = FALSE, row.names = FALSE)
```

## E4 PLX5622 male vs E4 control male within each cluster
```{r E4PM_vs_E4CM, warning=FALSE, message=FALSE, eval=FALSE}
# intitialize variables
cell_types <- levels(mouse.annotated$annotated_clusters)
#cell_types <- cell_types[!cell_types %in% c("Neurons","Pericytes")]
e4.male.df <- data.frame()

# loop through clusters
for (i in cell_types) {
  print(i)
  cluster <- subset(mouse.annotated, annotated_clusters == i)
  Idents(cluster) <- cluster$group
  markers <- FindMarkers(object = cluster,
                         ident.1 = "E4PM",
                         ident.2 = "E4CM",
                         only.pos = FALSE, # default
                         min.pct = 0.10,  # default
                         test.use = "MAST",
                         verbose = TRUE,
                         assay = "RNA")
  if(nrow(markers) == 0) {
    next
  }
  markers$cluster <- i
  markers$gene <- rownames(markers)
  e4.male.df <- rbind(e4.male.df, markers)
}

# reformat table
colnames(e4.male.df)[c(3,4)] <- c("percent_E4PM","percent_E4CM")
rownames(e4.male.df) <- 1:nrow(e4.male.df)
e4.male.df$percent_difference <- abs(e4.male.df$percent_E4PM - e4.male.df$percent_E4CM)
e4.male.df <- e4.male.df[,c(6,7,1,5,2,3,4,8)]

# write table
write.table(e4.male.df, "../../results/all_clusters_pass1/DEGs/E4PM_vs_E4CM_DEGs.tsv", 
            sep = "\t", quote = FALSE, row.names = FALSE)
```


## E3 PLX5622 female vs E3 control female within each cluster
```{r E3PF_vs_E3CF, warning=FALSE, message=FALSE, eval=FALSE}
# intitialize variables
cell_types <- levels(mouse.annotated$annotated_clusters)
e3.female.df <- data.frame()

# loop through clusters
for (i in cell_types) {
  print(i)
  cluster <- subset(mouse.annotated, annotated_clusters == i)
  Idents(cluster) <- cluster$group
  markers <- FindMarkers(object = cluster,
                         ident.1 = "E3PF",
                         ident.2 = "E3CF",
                         only.pos = FALSE, # default
                         min.pct = 0.10,  # default
                         test.use = "MAST",
                         verbose = TRUE,
                         assay = "RNA")
  if(nrow(markers) == 0) {
    next
  }
  markers$cluster <- i
  markers$gene <- rownames(markers)
  e3.female.df <- rbind(e3.female.df, markers)
}

# reformat table
colnames(e3.female.df)[c(3,4)] <- c("percent_E3PF","percent_E3CF")
rownames(e3.female.df) <- 1:nrow(e3.female.df)
e3.female.df$percent_difference <- abs(e3.female.df$percent_E3PF - e3.female.df$percent_E3CF)
e3.female.df <- e3.female.df[,c(6,7,1,5,2,3,4,8)]

# write table
write.table(e3.female.df, "../../results/all_clusters_pass1/DEGs/E3PF_vs_E3CF_DEGs.tsv", 
            sep = "\t", quote = FALSE, row.names = FALSE)
```

## E4 PLX5622 female vs E4 control female within each cluster
```{r isoform_de, warning=FALSE, message=FALSE, eval=FALSE}
# intitialize variables
cell_types <- levels(mouse.annotated$annotated_clusters)
cell_types <- cell_types[!cell_types %in% "Neurons"]
e4.female.df <- data.frame()

# loop through clusters
for (i in cell_types) {
  print(i)
  cluster <- subset(mouse.annotated, annotated_clusters == i)
  Idents(cluster) <- cluster$group
  markers <- FindMarkers(object = cluster,
                         ident.1 = "E4PF",
                         ident.2 = "E4CF",
                         only.pos = FALSE, # default
                         min.pct = 0.10,  # default
                         test.use = "MAST",
                         verbose = TRUE,
                         assay = "RNA")
  if(nrow(markers) == 0) {
    next
  }
  markers$cluster <- i
  markers$gene <- rownames(markers)
  e4.female.df <- rbind(e4.female.df, markers)
}

# reformat table
colnames(e4.female.df)[c(3,4)] <- c("percent_E4PF","percent_E4CF")
rownames(e4.female.df) <- 1:nrow(e4.female.df)
e4.female.df$percent_difference <- abs(e4.female.df$percent_E4PF - e4.female.df$percent_E4CF)
e4.female.df <- e4.female.df[,c(6,7,1,5,2,3,4,8)]

# write table
write.table(e4.female.df, "../../results/all_clusters_pass1/DEGs/E4PF_vs_E4CF_DEGs.tsv", 
            sep = "\t", quote = FALSE, row.names = FALSE)
```

## Compare DEGs
```{r,eval=FALSE}
# read tables
female.df <- read.table(
  "../../results/all_clusters_pass1/DEGs/E4_F_vs_E3_F_DEGs.tsv",
  sep = "\t", header = TRUE)
male.df <- read.table(
  "../../results/all_clusters_pass1/DEGs/E4_M_vs_E3_M_DEGs.tsv",
  sep = "\t", header = TRUE)

# filter
female.df <- female.df[female.df$p_val_adj < 0.05,]
male.df <- male.df[male.df$p_val_adj < 0.05,]

# add columns
direction <- female.df$avg_log2FC > 0
direction <- gsub(TRUE, "E4_female_up", direction)
direction <- gsub(FALSE, "E4_female_down", direction)
female.df$direction <- direction
direction <- male.df$avg_log2FC > 0
direction <- gsub(TRUE, "E4_male_up", direction)
direction <- gsub(FALSE, "E4_male_down", direction)
male.df$direction <- direction

# reformat tables
female.df2 <- female.df %>%
  dplyr::count(cluster,direction) %>%
  tidyr::spread(cluster, n)
male.df2 <- male.df %>%
  dplyr::count(cluster,direction) %>%
  tidyr::spread(cluster, n)

# master table
df <- smartbind(female.df2, male.df2)
df
```

```{r,echo=FALSE,eval=FALSE}
# save
write.table(
  df, 
  "../../results/all_clusters_pass1/DEGs/DEG_comparison_pvaladj_0.05.tsv",
  sep = "\t", quote = FALSE, row.names = FALSE)
```

## Upset plot
```{r upset_plots,eval=FALSE}
# read tables
female.df <- read.table("../../results/all_clusters_pass1/DEGs/E4_F_vs_E3_F_DEGs.tsv",
                         sep = "\t", header = TRUE)
male.df <- read.table("../../results/all_clusters_pass1/DEGs/E4_M_vs_E3_M_DEGs.tsv",
                         sep = "\t", header = TRUE)

# filter tables
female.df <- female.df[female.df$p_val_adj < 0.05,]
male.df <- male.df[male.df$p_val_adj < 0.05,]

clusters <- levels(mouse.annotated$annotated_clusters)
clusters <- cell_types[1:13]
for (i in clusters) {
  # Subset df by cluster
  female <- subset(female.df, female.df$cluster == i)
  sex <- subset(male.df, male.df$cluster == i)
  
  # Subset lists
  female_up <- subset(female$gene, female$avg_log2FC > 0)
  female_down <- subset(female$gene, female$avg_log2FC < 0)
  male_up <- subset(sex$gene, sex$avg_log2FC > 0)
  male_down <- subset(sex$gene, sex$avg_log2FC < 0)
  list_input <- list("E4 Female Up-regulated" = female_up,
                     "E4 Male Up-regulated" = male_up,
                     "E4 Female Down-regulated" = female_down,
                     "E4 Male Down-regulated" = male_down)
  data <- fromList(list_input)
  
  # store names
  names <- c("E4 Male Down-regulated","E4 Female Down-regulated",
             "E4 Male Up-regulated","E4 Female Up-regulated")
  
  # plot
  upset_gene <- ComplexUpset::upset(data, 
                      names,
                      set_sizes=(
                        upset_set_size()
                        + geom_text(aes(label=..count..), hjust=1.1, stat='count')
                        + expand_limits(y=200)),
                      queries = list(upset_query("E4 Female Up-regulated", fill = "red"),
                                     upset_query("E4 Male Up-regulated", fill = "red"),
                                     upset_query("E4 Female Down-regulated", fill = "blue"),
                                     upset_query("E4 Male Down-regulated", fill = "blue")),
                      base_annotations = list('Intersection size' = (
                        intersection_size(bar_number_threshold=1, width=0.5)
                        + scale_y_continuous(expand=expansion(mult=c(0, 0.05)),limits = c(0,150)) # space on top
                        + theme(
                              # hide grid lines
                              panel.grid.major=element_blank(),
                              panel.grid.minor=element_blank(),
                              # show axis lines
                              axis.line=element_line(colour='black')))),
                      stripes = upset_stripes(
                        geom=geom_segment(size=12),  # make the stripes larger
                        colors=c('grey95', 'white')),
                      # to prevent connectors from getting the colorured
                      # use `fill` instead of `color`, together with `shape='circle filled'`
                      matrix = intersection_matrix(
                        geom=geom_point(
                          shape='circle filled',
                          size=3,
                          stroke=0.45)),
                      sort_sets=FALSE,
                      sort_intersections='descending'
                    )
  upset_gene <- upset_gene + ggtitle(paste0(i,", adj_p_val < 0.05"))
  i <- gsub(" ","_",i)
  i <- gsub("/","_",i)
  i <- gsub("-","_",i)
  pdf(paste0("../../results/all_clusters_pass1/upset/upset_",tolower(i),".pdf"), height = 6, width = 8)
  print(upset_gene)
  dev.off()
}
```

## Volcano
```{r volcano_plot, message=FALSE, eval=FALSE}
variables <- c("E4_F_vs_E3_F","E4_M_vs_E3_M")
all_clusters <- levels(mouse.annotated$annotated_clusters)
all_clusters <- all_clusters[1:13]

for (i in variables) {
  
  # read DEG file
  if (i == "E4_F_vs_E3_F") {
    treatment_vs_control <- 
      read.delim("../../results/all_clusters_pass1/DEGs/E4_F_vs_E3_F_DEGs.tsv",
                 sep = "\t")
  } else {
    treatment_vs_control <-
      read.delim("../../results/all_clusters_pass1/DEGs/E4_M_vs_E3_M_DEGs.tsv",
                 sep = "\t")
  }
  
  # assign colors
  color_values <- vector()
  max <- nrow(treatment_vs_control)
  for(row in 1:max){
    if (treatment_vs_control$p_val_adj[row] < 0.05){
      if (treatment_vs_control$avg_log2FC [row] > 0){
        color_values <- c(color_values, 1) # 1 when logFC > 0 and FDRq < 0.05
      }
      else if (treatment_vs_control$avg_log2FC[row] < 0){
        color_values <- c(color_values, 2) # 2 when logFC < 0 and FDRq < 0.05
      }
    }
    else{
      color_values <- c(color_values, 3) # 3 when FDRq >= 0.05
    }
  }
  treatment_vs_control$color_adjpval_0.05 <- factor(color_values)
  
  # loop through clusters
  for (j in all_clusters) {
    
    # subset cluster
    data <- subset(treatment_vs_control, cluster == j)
    
    # plot only if there are DEGs with p_val_adj < 0.05
    num <- subset(data, p_val_adj < 0.05)
    num <- nrow(num)
    if(num != 0) {
        
      # subset genes to label
      up <- data[data$color_adjpval_0.05 == 1,]
      up10 <- up[1:10,]
      down <- data[data$color_adjpval_0.05 == 2,]
      down10 <- down[1:10,]
      
      # set manual colors
      if (!1 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("blue","gray")
      } else if (!2 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("red","gray")
      } else if (!1 %in% unique(data$color_adjpval_0.05) && !2 %in% unique(data$color_adjpval_0.05)) {
        my_colors <- c("gray")
      } else {
        my_colors <- c("red","blue","gray")
      }
      
      # set significance threshold
      hadjpval <- (-log10(max(
        data$p_val[data$p_val_adj < 0.05], 
        na.rm=TRUE)))

      # plot
      p <-
        ggplot(data = data, 
               aes(x = avg_log2FC,  # x-axis is logFC
                   y = -log10(p_val),  # y-axis will be -log10 of P.Value
                   color = color_adjpval_0.05)) +  # color is based on factored color column
        geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
        theme_bw() +  # set color theme
        theme(legend.position = "none") +  # no legend
        scale_color_manual(values = my_colors) +  # set factor colors
        labs(
          title = "", # no main title
          x = expression(log[2](FC)), # x-axis title
          y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
        ) +
        theme(axis.title.x = element_text(size = 10),
              axis.text.x = element_text(size = 10)) +
        theme(axis.title.y = element_text(size = 10),
              axis.text.y = element_text(size = 10)) +
        geom_hline(yintercept = hadjpval,  #  horizontal line
                           colour = "#000000",
                           linetype = "dashed") +
        ggtitle(paste0(j,"\n",i,", p_val_adj < 0.05")) +
        geom_text_repel(data = up10,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "maroon", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        ) +
        geom_text_repel(data = down10,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "navyblue", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        )
      p
      
      # save
      clus <- j
      clus <- gsub(" ","_",clus)
      clus <- gsub("/","_",clus)
      clus <- gsub("-","_",clus)
      if (i == "E4_F_vs_E3_F") {
        path <- paste0("../../results/all_clusters_pass1/volcano/E4_F_vs_E3_F/",
                       tolower(clus),"_female_volcano")
      } else {
        path <- paste0("../../results/all_clusters_pass1/volcano/E4_M_vs_E3_M/",
                       tolower(clus),"_male_volcano")      
      }
      pdf(paste(path,".pdf"), height = 8, width = 8)
      print(p)
      dev.off()
      
      print(paste("i =",i,", j =",j))
      
    }
  } # end loop through clusters
} # end loop through variables
```

## Metascape input
```{r}
both_up <- lax_up[lax_up %in% strict_up]
length(both_up)
write.table(as.data.frame(both_up), 
            "../../results/meninges_comparison/group_sex_SNAP25_FDRq_0.05_shared_upreg.tsv",
            quote = FALSE, row.names = FALSE, col.names = FALSE)

lax_up_unique <- lax_up[!lax_up %in% strict_up]
length(lax_up_unique)
write.table(as.data.frame(lax_up_unique), 
            "../../results/meninges_comparison/group_sex_SNAP25_FDRq_0.05_lax_unique_upreg.tsv",
            quote = FALSE, row.names = FALSE, col.names = FALSE)
```
